src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h

```diff
@@ -395,13 +466,25 @@ namespace OpenMS
           {
             if (Int(it_rt_2->getMSLevel()) == ms_level)
             {
-              double weight = 1;
-              if (average_type == "gaussian")
+              bool add = true;
+              // if precursor_mass_ppm >=0, two spectra should have the same mass. otherwise it_rt_2 is skipped.
+              if (precursor_mass_ppm >= 0 && ms_level >= 2 && it_rt->getPrecursors().size() > 0 &&
+                  it_rt_2->getPrecursors().size() > 0)
+              {
+                double mz1 = it_rt->getPrecursors()[0].getMZ();
+                double mz2 = it_rt_2->getPrecursors()[0].getMZ();
+                add = areMassesMatched(mz1, mz2, precursor_mass_ppm, precursor_max_charge);
+              }
+              if (add)
               {
-                weight = std::exp(factor * pow(it_rt_2->getRT() - it_rt->getRT(), 2));
+                double weight = 1;
+                if (average_type == "gaussian")
+                {
+                  weight = std::exp(factor * pow(it_rt_2->getRT() - it_rt->getRT(), 2));
+                }
+                std::pair<Size, double> p(m, weight);
+                spectra_to_average_over[n].push_back(p);
```

Comment:
if you feel like it, you can shorten this to
`spectra_to_average_over[n].emplace_back(m, weight);`

Reply:
This type of sophisticated use is something I need to learn. Many thanks. Will try.